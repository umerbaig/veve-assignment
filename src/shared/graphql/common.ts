import { Field, ObjectType, InputType } from '@nestjs/graphql';

@ObjectType()
export class PageInfo {
  @Field(type => Number)
  limit: number;

  @Field(type => Number)
  skip: number;

  @Field(type => Number)
  total: number;

  @Field(type => Boolean)
  has_more: boolean;
}

@InputType()
export class Pagination {
  @Field(type => Number)
  limit: number;

  @Field(type => Number)
  skip: number;
}
