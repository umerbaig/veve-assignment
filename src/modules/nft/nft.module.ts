import { Logger, Module } from '@nestjs/common';

import { NftResolver } from './graphql/nft.resolver';
import { TypeOrmModule } from '@nestjs/typeorm';
import { NftModel } from '../../models/nft.model';
import { NftService } from './nft.service';

@Module({
  imports: [TypeOrmModule.forFeature([NftModel])],
  providers: [NftResolver, NftService, Logger],
  exports: [TypeOrmModule],
})
export class NftModule {}
