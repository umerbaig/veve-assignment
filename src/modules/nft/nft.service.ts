import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { NftModel } from '../../models/nft.model';
import { hasMoreRec } from '../../utils/index';
import { NftsResp, Nft } from './graphql/nft.type';
import {
  InternalServerError,
  NotFoundError,
} from '../../shared/errors/errors.messages';

@Injectable()
export class NftService {
  constructor(
    @InjectRepository(NftModel)
    private nftRepository: Repository<NftModel>,
    private logger: Logger,
  ) {}

  async findAll(
    limit: number,
    skip: number,
    userId: number,
  ): Promise<NftsResp> {
    try {
      const [results, total] = await this.nftRepository.findAndCount({
        where: { owner: userId },
        take: limit,
        skip,
      });

      return {
        page_info: {
          limit,
          skip,
          total,
          has_more: hasMoreRec(total, skip, limit),
        },
        edges: results,
      };
    } catch (err) {
      throw InternalServerError('Something went wrong');
    }
  }

  async transferNft(
    id: number,
    currentUserId: number,
    newUserId: number,
  ): Promise<Nft> {
    const nft = await this.nftRepository.findOne({
      where: { owner: currentUserId, id },
    });

    if (!nft) {
      throw NotFoundError('Nft not found!');
    }

    const updatedNft = await this.nftRepository.update(id, {
      owner: newUserId,
    });

    if (!updatedNft) {
      throw InternalServerError('Something went wrong');
    }

    return this.nftRepository.findOne({ where: { id } });
  }
}
