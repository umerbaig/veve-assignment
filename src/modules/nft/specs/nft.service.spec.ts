import { Logger } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { NftModel } from '../../../models/nft.model';
import { NftService } from '../nft.service';
import { nft, nfts } from './nft.seed';
import { Repository } from 'typeorm';
import { Promise } from 'bluebird';
import * as TypeMoq from 'typemoq';
import to from 'await-to-js';

describe('NftService', () => {
  let service: NftService;
  const nftModelMock = TypeMoq.Mock.ofType(Repository<NftModel>);

  beforeEach(async () => {
    nftModelMock
      .setup((x) => x.findAndCount())
      .returns(() => Promise.resolve(nfts as NftModel[]));

    const module = await Test.createTestingModule({
      imports: [],
      providers: [
        {
          provide: getRepositoryToken(NftModel),
          useValue: nftModelMock.target,
        },
        NftService,
        Logger,
      ],
    }).compile();
    service = await module.get<NftService>(NftService);
  });

  afterEach(() => {
    nftModelMock.reset();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
    expect.assertions(1);
  });

  describe('findAll', () => {
    it('should return all nfts by user id', async () => {
      nftModelMock
        .setup((x) => x.findAndCount())
        .returns(() => Promise.resolve(nfts as NftModel[]));

      const [err, data] = await to(service.findAll(10, 0, 2));
      expect(data).toMatchSnapshot();
    });

    it('should throw error that something went wrong', async () => {
      nftModelMock
        .setup((x) => x.findAndCount())
        .returns(() => Promise.reject(nfts as NftModel[]));

      const [err, data] = await to(service.findAll(10, 0, 2));
      expect(err).toMatchSnapshot();
    });
  });

  describe('transferNft', () => {
    it('should transfer nft to the new owner', async () => {
      nftModelMock
        .setup((x) => x.findOne({ where: { id: 1 } }))
        .returns(() => Promise.resolve(nft as NftModel));
      nftModelMock
        .setup((x) => x.update(1, { owner: 2 }))
        .returns(() => Promise.resolve(nft as NftModel));

      const [err, data] = await to(service.transferNft(1, 2, 3));
      expect(data).toMatchSnapshot();
    });

    it('should throw error that nft not found', async () => {
      nftModelMock
        .setup((x) => x.findOne({ where: { id: 9 } }))
        .returns(() => Promise.reject());

      const [err, data] = await to(service.transferNft(5, 2, 3));
      expect(err).toMatchSnapshot();
    });
  });
});
