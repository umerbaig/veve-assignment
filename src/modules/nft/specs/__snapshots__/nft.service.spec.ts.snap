// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NftService findAll should return all nfts by user id 1`] = `
Object {
  "edges": Object {
    "blockchainLink": "test",
    "description": "test",
    "imageUrl": "test",
    "minDate": 2023-06-21T20:24:07.780Z,
    "name": "test",
  },
  "page_info": Object {
    "has_more": false,
    "limit": 10,
    "skip": 0,
    "total": undefined,
  },
}
`;

exports[`NftService findAll should throw error that something went wrong 1`] = `[Error: Something went wrong]`;

exports[`NftService transferNft should throw error that nft not found 1`] = `undefined`;

exports[`NftService transferNft should transfer nft to the new owner 1`] = `
Object {
  "blockchainLink": "test",
  "description": "test",
  "imageUrl": "test",
  "minDate": 2023-06-21T20:24:07.780Z,
  "name": "test",
}
`;
